Add advecting sinusoidal density soln for Newtonian Euler

nilsdeppe opened the pull request:
## Proposed changes

A simple sinusoidal density profiling advecting through the domain.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

fmahebert commented on src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Sinusoid.hpp:
```diff
@@ -0,0 +1,121 @@
+ *
+ * \f{align*}{
+ *  \sum_{i} v^i = 1
+ * \f}
```
For my understanding, why this contraint?

nilsdeppe replied:
If I remember correctly it's only a solution if this is satisfied. It's been a while since I plugged it all in and did the math, but that's what I got out :)

fmahebert replied:
I agree... I didn't work the solution out fully, but just enough to see how this is needed from balancing the d(rho)/dt with the div(flux(rho)) terms 👍

kidder reviewed (changes requested):
This is just a restriction of the RelativisticEuler::SmoothFlow solution 
```
rho = 1 + A sin[k_i x^i - k_i v^i t]
``` 
which is also valid for NewtonianEuler.  Just reuse that solution which is valid for arbitrary wave direction k_i, and arbitrary constants for the velocity, pressure, and adiabatic index, and for amplitudes of the sinusoid A between [-1,1]...   (Note the solution is actually valid for an arbitrary profile F[k_i x^i - k_i v^i t], though physical only for a non-negative profile...)

nilsdeppe commented:
I don't understand what you want at all...

nilsdeppe commented:
Are you saying make this dependent on RelEuler? Because I would disagree with that... If you're asking me to generalize this, my response is: it's not clear that's worth the effort

kidder commented:
GrMhd::Solutions::SmoothFlow is already implemented in terms of RelativisticEuler::Solutions::SmoothFlow.  I don't see how this is any different.

nilsdeppe commented:
After a telecon with @kidder we figured out what he was trying to ask is: "Can we have a unified solution that is used in NE, RE, and GRMHD?" The result is that I will see if I can unify the smooth flow solutions (advecting sinusoid) into a `AnalyticSolutions/CommonHydro` directory and have the different implementations wrap that, similar to how the GRMHD smooth flow solution wraps the RE smooth flow solution

nilsdeppe committed:
Add AnalyticSolutions/Hydro/SmoothFlow.?pp

Add generic smooth flow solution that can be shared between relativistic and
Newtonian Euler.

nilsdeppe committed:
Fix RelativisticEuler::SmoothFlow documentation

The math just didn't render at all.

nilsdeppe committed:
Use Hydro::SmoothFlow in RelEuler::SmoothFlow

nilsdeppe committed:
Simplify GrMhd::SmoothFlow test

nilsdeppe committed:
Add NewtonianEuler::SmoothFlow solution

nilsdeppe committed:
Add NewtonianEulerSmoothFlow?D executables

nilsdeppe commented:
Well this exploded a little bit... But I did the following:
- copy RelEuler:SmoothFlow to Hydro/SmoothFlow and generalize to support non-relativistic systems. This involved 1) add the template parameter 2) disable Lorentz factor when not relativistic 3) get calculation of specific enthalpy correct. Other than that the code is a copy-paste from RelEuler. I give you this detail to make review easier since 98% of these files have been reviewed. Please do look at the test, though.
- Have RelEuler use the new general hydro solution. This basically just deletes code...
- Add NE smooth flow solution.

I also did some minor cleanup of the python files for the tests.

kidder commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Hydro/Test_SmoothFlow.cpp:
```diff
@@ -0,0 +1,144 @@
+                  "[Unit][PointwiseFunctions]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "PointwiseFunctions/AnalyticSolutions/Hydro"};
+  // test<true>();
```
why is this commented out?

kidder reviewed (changes requested):
also a clang-tidy error

nilsdeppe commented:
I've pushed (and squashed) enabling the test and fixing clang-tidy since they were both boring trivial changes

kidder commented:
something went wrong with python in one of the builds...

nilsdeppe commented:
Yea, seems to be python2 specific... I'll investigate today to see what's happening :)

nilsdeppe commented:
I've fixed the python2 build issues. I needed to add `__init__.py` to `GrMhd` and `Hydro` test directories. I squashed this in directly since it seemed a minor and detail change

kidder commented:
@fmahebert please look at

fmahebert commented on src/PointwiseFunctions/AnalyticSolutions/Hydro/SmoothFlow.hpp:
```diff
@@ -0,0 +1,187 @@
+  }
+
+ private:
+  template <size_t SpatialDim, bool LocalIsRelativistic>
```
SpatialDim -> LocalDim

fmahebert commented on src/PointwiseFunctions/AnalyticSolutions/Hydro/Solutions.hpp:
```diff
@@ -0,0 +1,17 @@
+ *
+ * For example, an advecting sinusoidal density profile is a solution to
+ * Newtonian Euler, relativistic Euler, and GRMHD, so the common parts of the
+ * solutions is implemented here.
```
solutions -> solution

fmahebert commented on src/PointwiseFunctions/AnalyticSolutions/Hydro/Solutions.hpp:
```diff
@@ -0,0 +1,17 @@
+namespace hydro {
+/*!
+ * \ingroup AnalyticSolutionsGroup
+ * \brief Holds classes implementing common portions of solution to various
```
solution -> solutions

fmahebert commented on src/PointwiseFunctions/AnalyticSolutions/Hydro/SmoothFlow.cpp:
```diff
@@ -0,0 +1,222 @@
+    :  // clang-tidy: do not std::move trivial types.
+      mean_velocity_(std::move(mean_velocity)),  // NOLINT
+      // clang-tidy: do not std::move trivial types.
+      wavevector_(std::move(wavevector)),  // NOLINT
```
Help: is it not true that std::moving a const-ref does nothing? Do I misremember my C++?

nilsdeppe replied:
yea, hence the `NOLINT`.... I've been meaning to clean this sort of thing up and clearly overlooked it here. I shall fix :)

fmahebert commented on src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/SmoothFlow.cpp:
```diff
@@ -0,0 +1,74 @@
+#undef TAG
+#undef INSTANTIATE_CLASS
+#undef INSTANTIATE_SCALARS
+#undef INSTANTIATE_VECTORS
```
Clean up macros: I think you can delete DTYPE, TAG, INSTANTIATE_SCALARS, INSTANTIATE_VECTORS

fmahebert commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/SmoothFlow.cpp:
```diff
@@ -167,29 +66,6 @@ bool operator!=(const SmoothFlow<Dim>& lhs,
-
 #undef DIM
 #undef DTYPE
 #undef TAG
```
Clean up macros here too

fmahebert reviewed (commented):
I don't see you using the new generic SmoothFlow for the GrMhd solution. Does this occur transitively by using the generic SmoothFlow in the relativistic Euler solution?

nilsdeppe commented:
Yes, that's right. Since RelEuler uses SmoothFlow, so does GRMHD. Should I add a comment to the Doxygen at the end? Maybe a `\note that this uses Hydro/SmoothFlow.?pp via the RelativisticEuler::SmothFlow class`?

fmahebert commented:
I don't think that the comment is needed (feel free to add it if you like, of course). I was asking from the point of view of checking that the PR was making additions/changes to all three hydro system solutions and only seeing two. But I now understand why 🙂

nilsdeppe commented:
Ah okay, sounds good :) Thank you for checking!

nilsdeppe commented:
Rebased on develop and pushed fixups. Thanks for the reviews :)

kidder reviewed (commented):
You have my permission to squash fixups...

fmahebert reviewed (commented):
LGTM, please squash

nilsdeppe commented:
Okay, rebased on develop (again) and squashed. Thank you for the great feedback :D
